src/lein_monolith/plugin.clj

```diff
@@ -82,16 +99,70 @@
 (defn build-inherited-profiles
   "Returns a map from profile keys to inherited profile maps."
   [monolith subproject]
-  (let [inherit-profile (inherited-profile
-                          monolith :inherit
-                          (:monolith/inherit subproject))
-        leaky-profile (inherited-profile
-                        monolith :inherit-leaky
-                        (:monolith/leaky subproject (boolean (:monolith/inherit subproject))))]
-    (cond-> nil
-      inherit-profile (assoc :monolith/inherited inherit-profile)
-      leaky-profile (assoc :monolith/leaky (vary-meta leaky-profile assoc :leaky true)))))
+  (reduce
+    (fn [acc [profile-key {:keys [types ks]}]]
+      (let [profile (some-> (choose-inheritance-source monolith types)
+                            (inherited-profile subproject ks)
+                            (maybe-mark-leaky types))]
+        (cond-> acc
+          profile (assoc profile-key profile))))
+    nil
+    profile-config))
+
+
+(def ^:private init-lock
+  "An object to lock on to ensure that projects are not initialized
+  concurrently. This prevents the mysterious 'unbound fn' errors that sometimes
+  crop up during parallel execution."
+  (Object.))
+
+
+(declare add-middleware)
+
+
+(defn init-subproject
+  "Reads and fully initializes a subproject with inherited monolith profiles."
+  [subproject]
+  (binding [eval/*dir* (:root subproject)]
+    (config/debug-profile "init-subproject"
+      (locking init-lock
+        (project/init-project
+          (add-middleware subproject))))))
+
 
+(def ^:private path-keys
+  "Project map keys for (re)source and test paths."
+  #{:resource-paths :source-paths :test-paths})
+
+
+(defn- add-profile-paths
+  "Update a profile paths entry by adding the paths from the given project.
+  Returns the updated profile."
+  [project profile k]
+  (update profile k (fn combine-colls
+                      [coll]
+                      (-> coll
+                          set
+                          (into (get project k))
+                          (vary-meta assoc :replace true)))))
+
+
+(defn merged-profile
+  "Constructs a profile map containing merged (re)source and test paths."
+  [monolith subprojects]
+  (let [profile
+        (reduce-kv
+          (fn [profile _project-name subproject]
+            (reduce (->> subproject
+                         (init-subproject)
```

Comment:
Ah, is this because you want to inherit the paths from the metaproject, canonicalize them, _then_ combine them into the merged path map for the `with-all` task?  I'm still 🙅‍♂️ on having a project init step here; it should be possible to do this without having to load everything.

I guess previously we were missing out on any path changes that would have been inserted by _other_ plugins, but that doesn't seem like it's been hurting anyone.

Reply:
Good point. I think that this should just absolutize the paths itself.